dev-docs/RFCs/v6.0/data-filter-rfc.md

```diff
@@ -0,0 +1,100 @@
+
+## Background
+
+A common use case in data visualization applications is to update deck.gl layers with a data filter. A data filter shows/hides certain objects in the data array by some dynamic criteria. For example, of an input array of taxi pick up locations, only show those within a given time window. The app may also need to change this filter frequently, e.g. show an animation of how pick up locations change over the course of a day.
+
+Currently, there are two ways to apply a data filer:
+
+* Supply the filtered data array to the layer's `data` prop. This triggers recalculation of all attributes, and can be expensive depending on the size of the data.
+* Update a single attribute to produce invisible elements. For example, return `0` with `getRadius` for the ScatterplotLayer. This requires a custom logic for each layer, and for some layers there is no accessor with this effect (e.g. PolygonLayer - transparent color does not hide an object from picking).
+
+
+## Proposed Features
+
+### New Layer Props
+
+Add two new props to base Layer:
+
+* `getFilterValue` (function|number) - accessor to the "filterable" value of each data object. Default `1`.
+* `filterRange` ([number, number]) - the [min, max] bounds of the filter values to display. Default `[0.5, 1.5]`.
+
+Add a default attribute to the base Layer attributeManager:
+
+* `instanceFilterValue` - auto populated with accessor `getFilterValue`.
+
+If neither props are supplied, the performance impact is minimal (1 generic vertex attribute + 1 uniform).
+
+#### Examples
+
+Use case 1: filter layer data by a moving time window, without re-generating any attribute:
+
+```js
+new ScatterplotLayer({
+    ...
+    getFilterValue: d => d.time,
+    filterRange: [currentTime, currentTime + 1000] // currentTime is a number updated per animation frame
+});
+```
+
+Use case 2: custom filtering without manipulating the data object, thus avoiding re-generating all attributes:
+
+```js
+new ScatterplotLayer({
+    ...
+    getFilterValue: d => isVisible(d) ? 1 : 0
+});
+```
+
+#### TODO
+
+* Add feature to Layer class
+* Add data filter prop forwarding to CompositeLayer
+* Add data filter prop handling to aggregation layers
+
+
+### New Shader Module
+
+Add a new shader module `filter` and add it as a default module. It adds the uniform `vec2 filterRange` from the layer prop with the same name. It also adds two vertex shader functions:
+
+* `bool filter_setVisibility(float filterValue)` - determine visibility by checking value against `filterRange`
+* `bool filter_setVisibility(bool visible)` - override data filter with custom logic
+
+Ands one fragment shader function:
+
+* `bool filter_isVisible()` - returns `false` if fragment should be discarded
+
+#### Examples
+
+To use this module, a layer may add the following to its vertex shader:
+
+```glsl
+attribute float instanceFilterValue;
+
+void main() {
+    if (not(filter_setVisibility(instanceFilterValue))) {
```

Comment:
What happens if we don't write to `gl_Position` in a vertex shader? [Ref](https://bugzilla.mozilla.org/show_bug.cgi?id=816328)

Reply:
I have tried this in Chrome and it works fine. I can see the behavior being device/browser dependent though.